presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionTreeUtils.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.analyzer;
+
+import com.facebook.presto.metadata.FunctionRegistry;
+import com.facebook.presto.sql.planner.optimizations.Predicates;
+import com.facebook.presto.sql.tree.DefaultExpressionTraversalVisitor;
+import com.facebook.presto.sql.tree.Expression;
+import com.facebook.presto.sql.tree.FunctionCall;
+import com.facebook.presto.sql.tree.Node;
+import com.google.common.collect.ImmutableList;
+
+import java.util.List;
+import java.util.function.Predicate;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.util.Objects.requireNonNull;
+
+final class ExpressionTreeUtils
+{
+    private ExpressionTreeUtils() {}
+
+    static List<FunctionCall> extractAggregateFunctions(Iterable<? extends Node> nodes, FunctionRegistry functionRegistry)
+    {
+        return extractExpressionsOfTypeUsingPredicate(nodes, FunctionCall.class, isAggregationPredicate(functionRegistry));
+    }
+
+    static List<FunctionCall> extractWindowFunctions(Iterable<? extends Node> nodes)
+    {
+        return extractExpressionsOfTypeUsingPredicate(nodes, FunctionCall.class, ExpressionTreeUtils::isWindowFunction);
+    }
+
+    static <T extends Expression> List<T> extractExpressionsOfType(
+            Iterable<? extends Node> nodes,
+            Class<T> clazz)
+    {
+        return extractExpressionsOfTypeUsingPredicate(nodes, clazz, Predicates.alwaysTrue());
+    }
+
+    private static Predicate<FunctionCall> isAggregationPredicate(FunctionRegistry functionRegistry)
+    {
+        return ((functionCall) -> (functionRegistry.isAggregationFunction(functionCall.getName()) || functionCall.getFilter().isPresent()) && !functionCall.getWindow().isPresent());
+    }
+
+    private static boolean isWindowFunction(FunctionCall functionCall)
+    {
+        return functionCall.getWindow().isPresent();
+    }
+
+    private static <T extends Expression> List<T> extractExpressionsOfTypeUsingPredicate(
```

Comment:
I spoke to Martin offline and we agreed to keep it as is, except for renaming the method.